src/util/file_log.rs

```diff
@@ -0,0 +1,228 @@
+    };
+    let duration = time::Duration::from_std(Duration::new(ONE_DAY_SECONDS, 0)).unwrap();
+    (day_start_tm.to_utc() + duration).to_local()
+}
+
+pub fn file_exists(file: &str) -> bool {
+    let path = Path::new(file);
+    path.exists() && path.is_file()
+}
+
+fn open_log_file(path: &str) -> io::Result<File> {
+    OpenOptions::new()
+        .append(true)
+        .create(true)
+        .open(path)
+}
+
+struct RotatingFileLoggerCore {
+    rollover_time: Tm,
+    file_path: String,
+    file: File,
+}
+
+impl RotatingFileLoggerCore {
+    fn new(path: &str) -> io::Result<RotatingFileLoggerCore> {
+        let file = try!(open_log_file(path));
+        let file_attr = fs::metadata(path).unwrap();
+        let file_modified_time = file_attr.modified().unwrap();
+        let rollover_time = compute_rollover_time(systemtime_to_tm(file_modified_time));
+        let ret = RotatingFileLoggerCore {
+            rollover_time: rollover_time,
+            file_path: path.to_string(),
+            file: file,
+        };
+        Ok(ret)
+    }
+
+    fn open(&mut self) {
+        self.file = open_log_file(&self.file_path).unwrap()
+    }
+
+    fn update_rollover_time(&mut self) {
+        let now = time::now();
+        self.rollover_time = compute_rollover_time(now);
+    }
+
+    fn close(&mut self) {
+        self.file.flush().unwrap()
+    }
+}
+
+pub fn init(level: &str, file_path: &str) -> Result<(), SetLoggerError> {
+    let l = logger::get_level_by_string(level);
+    log::set_logger(|max_log_level| {
+        max_log_level.set(l);
+        Box::new(RotatingFileLogger::new(level, file_path).unwrap())
+    })
+}
+
+/// A log implemetation which writes to file and rotates by day.
+pub struct RotatingFileLogger {
+    level: LogLevelFilter,
+    core: Arc<RwLock<RotatingFileLoggerCore>>,
+}
+
+impl RotatingFileLogger {
+    pub fn new(level: &str, file_path: &str) -> io::Result<RotatingFileLogger> {
+        let core = try!(RotatingFileLoggerCore::new(file_path));
+        let ret = RotatingFileLogger {
+            level: logger::get_level_by_string(level),
+            core: Arc::new(RwLock::new(core)),
+        };
+        Ok(ret)
+    }
+
+    fn should_rollover(&self) -> bool {
+        let core = self.core.rl();
+        time::now() > core.rollover_time
+    }
+
+    fn do_rollover(&self) {
+        let mut core = self.core.wl();
+        core.close();
+        let mut s = core.file_path.clone();
+        s.push_str(".");
+        s.push_str(&time::strftime("%Y%m%d", &core.rollover_time).unwrap());
+        fs::rename(&core.file_path, &s).unwrap();
+        core.update_rollover_time();
+        core.open()
+    }
+}
+
+
+impl Log for RotatingFileLogger {
+    fn enabled(&self, metadata: &LogMetadata) -> bool {
+        metadata.level() <= self.level
+    }
+
+    fn log(&self, record: &LogRecord) {
+        if self.enabled(record.metadata()) {
+            if self.should_rollover() {
+                self.do_rollover()
```

Comment:
multi threads can enter here at same time, so we will do do_rollover many times. 
you can put `let mut core = self.core.wl()` at first.